File: backend/app/schemas/review.py
"""坐标审阅契约：位置为不可变原文的 Unicode 码点 [start, end)。"""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.collaboration import CollaborationReport

MAX_ISSUES = 5000
MAX_REVIEW_BYTES = 8 * 1024 * 1024
MAX_VERSIONS = 20


class ReviewIssue(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    original: str = Field(min_length=1, max_length=100000)
    type: str = Field(min_length=1, max_length=64)
    suggestion: str = Field(max_length=100000)
    explanation: str | None = Field(default="", max_length=20000)
    severity: str = Field(max_length=32)
    chunk_index: int | None = Field(default=None, ge=0, strict=True)
    start: int | None = Field(default=None, ge=0, strict=True)
    end: int | None = Field(default=None, ge=0, strict=True)
    accepted: bool = Field(default=False, alias="_accepted", strict=True)
    ignored: bool = Field(default=False, alias="_ignored", strict=True)

    @model_validator(mode="after")
    def validate_flags(self):
        if self.accepted and self.ignored:
            raise ValueError("不能同时采纳和忽略同一问题")
        if (self.start is None) != (self.end is None):
            raise ValueError("start/end 必须同时提供或同时为空")
        if self.accepted and self.start is None:
            raise ValueError("未定位的问题不可采纳")
        if self.accepted and not self.suggestion and self.type != "sensitive":
            raise ValueError("普通修改的 suggestion 不可为空")
        return self


class FailedChunk(BaseModel):
    model_config = ConfigDict(extra="forbid")
    chunk_index: int = Field(ge=0, strict=True)
    start: int = Field(ge=0, strict=True)
    end: int = Field(ge=0, strict=True)
    text: str = Field(max_length=100000)
    error_code: str | None = Field(default=None, max_length=200)


class ReviewCoverage(BaseModel):
    model_config = ConfigDict(extra="forbid")
    status: Literal["complete", "partial"]
    total_chunks: int = Field(ge=0, le=100000, strict=True)
    completed_chunks: int = Field(ge=0, le=100000, strict=True)
    failed_chunks: list[FailedChunk] = Field(default_factory=list, max_length=5000)

    @model_validator(mode="after")
    def validate_counts(self):
        if self.completed_chunks + len(self.failed_chunks) != self.total_chunks:
            raise ValueError("覆盖统计与失败分片数不一致")
        expected = "partial" if self.failed_chunks else "complete"
        if self.status != expected:
            raise ValueError("覆盖状态与失败分片不一致")
        indices = [chunk.chunk_index for chunk in self.failed_chunks]
        if len(set(indices)) != len(indices) or any(i >= self.total_chunks for i in indices):
            raise ValueError("失败分片序号重复或越界")
        return self


class ReviewCompareModel(BaseModel):
    model_config = ConfigDict(extra="forbid")

    config_id: int = Field(gt=0, strict=True)
    config_name: str
    model: str
    domain: str = "general"
    depth: Literal["quick", "standard", "deep"] = "standard"
    success: bool = Field(strict=True)
    error: str | None = None
    elapsed_ms: int = Field(default=0, ge=0, strict=True)
    issues: list[ReviewIssue] = Field(default_factory=list, max_length=MAX_ISSUES)
    coverage: ReviewCoverage | None = None

    @model_validator(mode="after")
    def validate_report_flags(self):
        if any(issue.accepted or issue.ignored for issue in self.issues):
            raise ValueError("逐模型报告不可保存采纳/忽略决策，请使用顶层 issues")
        return self


class ReviewCompareState(BaseModel):
    model_config = ConfigDict(extra="forbid")

    results: list[ReviewCompareModel] = Field(min_length=2, max_length=4)

    @model_validator(mode="after")
    def validate_models(self):
        ids = [item.config_id for item in self.results]
        if len(set(ids)) != len(ids):
            raise ValueError("对比模型 config_id 不可重复")
        return self


class ReviewWriteRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    revision: int = Field(ge=0, le=2147483646, strict=True)
    issues: list[ReviewIssue] = Field(max_length=MAX_ISSUES)
    coverage: ReviewCoverage | None = None
    compare: ReviewCompareState | None = None
    depth: Literal["quick", "standard", "deep"] | None = None
    config_id: int | None = Field(default=None, ge=1, strict=True)

    @model_validator(mode="after")
    def validate_size(self):
        if len(self.model_dump_json(by_alias=True).encode("utf-8")) > MAX_REVIEW_BYTES:
            raise ValueError("审阅 JSON 超过 8 MiB 限制")
        return self


class ReviewVersionRequest(ReviewWriteRequest):
    label: str | None = Field(default=None, max_length=200)


class ReviewExportRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    revision: int = Field(ge=0, strict=True)
    version_id: str | None = Field(default=None, min_length=1, max_length=100)
    format: Literal["docx", "txt"]


class DocumentReviewExportRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    issues: list[ReviewIssue] = Field(max_length=MAX_ISSUES)
    format: Literal["docx", "txt"]

    @model_validator(mode="after")
    def validate_size(self):
        if len(self.model_dump_json(by_alias=True).encode("utf-8")) > MAX_REVIEW_BYTES:
            raise ValueError("审阅 JSON 超过 8 MiB 限制")
        return self


class ReviewVersion(BaseModel):
    id: str
    label: str
    created_at: datetime
    issues: list[ReviewIssue]
    coverage: ReviewCoverage | None = None
    compare: ReviewCompareState | None = None
    modified_text: str


class ReviewResponse(BaseModel):
    record_id: int
    revision: int
    original_text: str
    source_file_id: str | None = None
    source_filename: str | None = None
    domain: str
    depth: str | None = None
    config_id: int | None = None
    issues: list[ReviewIssue]
    coverage: ReviewCoverage | None = None
    compare: ReviewCompareState | None = None
    collaboration: CollaborationReport | None = None
    modified_text: str
    versions: list[ReviewVersion] = Field(default_factory=list)

File: frontend/src/api/review.ts
/** 审阅草稿、显式版本和导出；不创建记录、不触发校对。 */
import request from '@/utils/request'
import type { ProofreadCoverage } from '@/api/proofread'
import type { CollaborationReport } from '@/api/collaboration'
import type { ReviewIssue } from '@/composables/useProofreadReview'
import { serializeCompareReview } from '@/utils/compareReview'

export interface ReviewCompareModel {
  config_id: number
  config_name: string
  model: string
  domain: string
  depth: string
  success: boolean
  error: string | null
  elapsed_ms: number
  issues: ReviewIssue[]
  coverage: ProofreadCoverage | null
}

export interface ReviewCompareState {
  results: ReviewCompareModel[]
}

export interface ReviewSnapshot {
  issues: ReviewIssue[]
  coverage?: ProofreadCoverage | null
  compare?: ReviewCompareState | null
  modified_text: string
}

export interface ReviewVersion extends ReviewSnapshot {
  id: string
  label: string
  created_at: string
}

export interface ReviewResponse extends ReviewSnapshot {
  record_id: number
  revision: number
  original_text: string
  source_file_id?: string | null
  source_filename?: string | null
  domain: string
  depth: string
  config_id?: number | null
  versions: ReviewVersion[]
  readonly collaboration?: CollaborationReport | null
}

/** restore 只回放本地审阅决策；接收方同时恢复 coverage，不自动写入服务端。 */
export interface ReviewRestorePayload {
  issues: ReviewIssue[]
  coverage: ProofreadCoverage | null
  compare?: ReviewCompareState | null
}

export interface SaveReviewPayload {
  revision: number
  issues: ReviewIssue[]
  coverage?: ProofreadCoverage | null
  compare?: ReviewCompareState | null
  depth?: string
  config_id?: number | null
}

export interface CreateReviewVersionPayload extends SaveReviewPayload {
  label?: string
}

export interface ReviewRequestOptions {
  signal?: AbortSignal
}

export interface ExportReviewPayload {
  revision: number
  version_id?: string
  format: 'docx' | 'txt'
}

interface ReviewError {
  message?: string
  response?: { status?: number; data?: unknown }
}

function detailText(detail: unknown): string {
  if (typeof detail === 'string') return detail
  if (Array.isArray(detail)) return detail.map(item => {
    if (item && typeof item === 'object' && 'msg' in item) return detailText(item.msg)
    return detailText(item)
  }).filter(Boolean).join('；')
  return ''
}

export function getReviewErrorDetail(error: unknown): string {
  const err = error as ReviewError | null
  const data = err?.response?.data
  const detail = data && typeof data === 'object' && 'detail' in data ? detailText(data.detail) : ''
  return detail || err?.message || '请求失败，请重试'
}

/** 保留原错误、HTTP 状态和拒绝语义，让调用方能明确处理 409/422。 */
async function reviewRequest<T>(promise: Promise<T>): Promise<T> {
  try {
    return await promise
  } catch (error) {
    const err = error as ReviewError | null
    const data = err?.response?.data
    if (typeof Blob !== 'undefined' && data instanceof Blob && err?.response) {
      try {
        err.response.data = JSON.parse(await data.text())
        err.message = getReviewErrorDetail(err)
      } catch {
        // 非 JSON 的错误响应仍抛出原错误，不把失败当成可下载的文件。
      }
    }
    throw error
  }
}

// 页面/组件显示错误（尤其 Blob detail），避免请求拦截器先弹出无意义的通用错误。
const silentHeaders = { 'X-Silent-Error': 'true' }

export function getReviewApi(recordId: number, options: ReviewRequestOptions = {}): Promise<ReviewResponse> {
  return reviewRequest(request.get<ReviewResponse, ReviewResponse>(`/history/${recordId}/review`, {
    ...options, headers: silentHeaders,
  }))
}

function wireIssues(issues: readonly ReviewIssue[]) {
  return issues.map(issue => ({
    original: issue.original,
    suggestion: issue.suggestion,
    type: issue.type,
    severity: issue.severity,
    explanation: issue.explanation,
    chunk_index: issue.chunk_index,
    start: issue.start == null || issue.start < 0 ? null : issue.start,
    end: issue.end == null || issue.end < 0 ? null : issue.end,
    _accepted: issue._accepted,
    _ignored: issue._ignored,
  }))
}

function wireReview(data: SaveReviewPayload) {
  return {
    revision: data.revision, issues: wireIssues(data.issues),
    ...('coverage' in data ? { coverage: data.coverage } : {}),
    ...('compare' in data ? { compare: serializeCompareReview(data.compare) } : {}),
    ...('depth' in data ? { depth: data.depth } : {}),
    ...('config_id' in data ? { config_id: data.config_id } : {}),
  }
}

export function saveReviewApi(recordId: number, data: SaveReviewPayload, options: ReviewRequestOptions = {}): Promise<ReviewResponse> {
  return reviewRequest(request.put<ReviewResponse, ReviewResponse>(`/history/${recordId}/review`, wireReview(data), {
    ...options, headers: silentHeaders,
  }))
}

export function createReviewVersionApi(recordId: number, data: CreateReviewVersionPayload, options: ReviewRequestOptions = {}): Promise<ReviewResponse> {
  return reviewRequest(request.post<ReviewResponse, ReviewResponse>(`/history/${recordId}/versions`, { ...wireReview(data), ...('label' in data ? { label: data.label } : {}) }, {
    ...options, headers: silentHeaders,
  }))
}

export function exportReviewApi(recordId: number, data: ExportReviewPayload, options: ReviewRequestOptions = {}): Promise<Blob> {
  return reviewRequest(request.post<Blob, Blob>(`/history/${recordId}/export`, data, {
    ...options, responseType: 'blob', headers: silentHeaders,
  }))
}

export function exportDocumentReviewApi(fileId: string, data: { issues: ReviewIssue[]; format: 'docx' | 'txt' }, options: ReviewRequestOptions = {}): Promise<Blob> {
  return reviewRequest(request.post<Blob, Blob>(`/document/${encodeURIComponent(fileId)}/export`, { ...data, issues: wireIssues(data.issues) }, {
    ...options, responseType: 'blob', headers: silentHeaders,
  }))
}
